lib/resty/http.lua

```diff
@@ -776,11 +776,65 @@ function _M.request_uri(self, uri, params)
     if not params.path then params.path = path end
     if not params.query then params.query = query end
 
-    local c, err = self:connect(host, port)
+    -- See if we should use a proxy to make this request
+    local proxy_host, proxy_port;
+    local proxy_uri = self:get_proxy_uri(scheme, host)
+    if proxy_uri then
+        local parsed_proxy_uri, err = self:parse_uri(proxy_uri, false)
+        if not parsed_proxy_uri then
+            return nil, err
+        end
+
+        proxy_host, proxy_port = parsed_proxy_uri[2], parsed_proxy_uri[3]
+    end
+
+    local c, err = self:connect(proxy_host or host, proxy_port or port)
     if not c then
         return nil, err
     end
 
+    if proxy_uri and scheme == "https" then
+        -- Make a CONNECT request to create a tunnel to the destination through
+        -- the proxy
+        local destination = host .. ":" .. port
+        local res, err = self:request({
+            method = "CONNECT",
+            path = destination,
+            headers = {
+                ["Host"] = destination
+            }
+        })
+
+        if not res then
+            return nil, err
+        end
+
+        if res.status < 200 or res.status > 299 then
+            return nil, "failed to establish a tunnel through a proxy: " .. res.status
+        end
+
+        -- don't keep this connection alive as the next request could target
+        -- any host and re-using the tunnel for that is not possible
+        self.keepalive = false
+    end
+
+    if proxy_uri and scheme == "http" then
+        -- http proxies expect to see the full URI in the request line
+        if port == 80 then
+            params.path = scheme .. "://" .. host .. path
```

Comment:
Isn't this the `uri` parameter? 

What I mean is is it necessary to create new string on every request? Wouldn't it be just fine to reuse the `uri` parameter passed to this function?